deep-learning-server/examples/aggregation.c

```diff
@@ -0,0 +1,235 @@
+                    axpy_cpu(l.n, 1, l.scales, 1, out.scales, 1);
+                    axpy_cpu(l.n, 1, l.rolling_mean, 1, out.rolling_mean, 1);
+                    axpy_cpu(l.n, 1, l.rolling_variance, 1, out.rolling_variance, 1);
+                }
+            }
+            if(l.type == CONNECTED){
+                axpy_cpu(l.outputs, 1, l.biases, 1, out.biases, 1);
+                axpy_cpu(l.outputs*l.inputs, 1, l.weights, 1, out.weights, 1);
+            }
+        }
+    }
+
+    // /No. of clients: for every layer, do division operation
+    n = n+1;
+    for(j = 0; j < net->n; ++j){
+        layer l = sum->layers[j];
+        if(l.type == CONVOLUTIONAL){
+            int num = l.n*l.c*l.size*l.size;
+            scal_cpu(l.n, 1./n, l.biases, 1);
+            scal_cpu(num, 1./n, l.weights, 1);
+                if(l.batch_normalize){
+                    scal_cpu(l.n, 1./n, l.scales, 1);
+                    scal_cpu(l.n, 1./n, l.rolling_mean, 1);
+                    scal_cpu(l.n, 1./n, l.rolling_variance, 1);
+                }
+        }
+        if(l.type == CONNECTED){
+            scal_cpu(l.outputs, 1./n, l.biases, 1);
+            scal_cpu(l.outputs*l.inputs, 1./n, l.weights, 1);
+        }
+    }
+
+    // save the weights of sum network
+    save_weights(sum, outfile);
+}
+
+
+// For ONNX format weights
+// this function reads all clients' weights file, load them into seperate 
+// networks, then from the first layer to the last layer, averages all 
+// weights if this layer has (i.e., Sum / No. of clients).
+//
+// - Input: all input arguments, including 1) network cfg 2) clients weight files
+// - Ouput: Aggregated weights file
+//
+void average_onnx(int argc, char *argv[])
+{
+    char *output_file_dir = argv[3];
+
+    // load the first weights file into sum network (as output later)
+    Onnx__ModelProto *model_sum = openOnnxFile(argv[4]);
+    if (model_sum != NULL){printf("model c1 loaded\n");}
+
+    // for all other weights files
+    int n = argc - 5;
+    for(int i = 0; i < n; ++i){
+        Onnx__ModelProto *model_c = openOnnxFile(argv[i+5]);
+        if (model_c != NULL){printf("model c%d loaded\n", i+2);}
+
+        // SUM: for each tensor in the graph
+        for(int j=0; j < model_sum->graph->n_initializer; j++){
+            int32_t dt = model_sum->graph->initializer[j]->data_type; //data type
+
+            //FLOAT or COMPLEX64
+            if (dt == 1 || dt == 14){
+                int ts = model_sum->graph->initializer[j]->n_float_data;
+                float *td_sum = model_sum->graph->initializer[j]->float_data;
+                float *td_c = model_c->graph->initializer[j]->float_data;
+                for(int k=0; k < ts; k++){
+                    td_sum[k] = td_sum[k] + td_c[k];
+                }
+                
+            // INT32, INT16, INT8, UINT16, UINT8, BOOL, or FLOAT16
+            }else if (dt==2 || dt==3 || dt==4 || dt==5 || dt==6 || dt==9 || dt==10){
```

Comment:
Bizarre spacing around `}else` here and elsewhere in this file.

Reply:
all of these are formatted